### DISCUSSION TITLE
Jest NextJS API routes

### DISCUSSION THREAD

**Discussion by rsheppard-dev** [vercel/next.js#41444] [Help]
I am struggling to find documentation on this. I am trying to implement tests in my api routes on Nextjs and am struggling.

I am using next-connect with Mongoose and Typescript.

Does anyone have a good tutorial or article on best way to do this?

**Comment by icyJoseph** [comment]
You could use `msw` to mock outgoing requests to 3rd parties, and `ts-mockito` to mock the inputs to the API:

```ts
// pseudo-code
import {setupServer} from 'msw/node';
import type {NextApiRequest} from 'next';
import {instance, mock, reset, when} from 'ts-mockito';

const server = setupServer(); // you'd pass handlers for 3rd party requests here

beforeAll(() => {
    server.listen();
});

afterAll(() => {
    server.close();
});

const mockedReq = mock<NextApiRequest>();

test('does not allow GET requests', async () => {
   when(mockedReq.method).thenReturn('GET');

   const req = instance(mockedReq);
   const res = new NextApiResponseMock(req);
   await statusHandler(req, res);
   
   expect(res.statusCode).toEqual(404);
});
```

Alternatively you could also use this library: https://www.npmjs.com/package/next-test-api-route-handler, which I've used for more complex situations where mocking via `ts-mockito` is not enough, or is not producing reliable tests.

**Comment by robreinhard** [comment]
What I ended up doing @rsheppard-dev was import each handler and mock the request and response object. Then, you can pass it through. This had issues with routes that contained custom configs though.
